### PR TITLE
[AutoDiff] [stdlib] Add JVP for 'differentiableReduce(_:_:)'.

### PR DESCRIPTION
**Pull request by rxwei**
Add a JVP for `differentiableReduce(_:_:)`.

### CODE REVIEW

**Comment by rxwei** [comment]
Note: This is currently missing unit tests.

File: stdlib/public/Differentiation/DifferentiationSupport.swift [hunk 1]
```diff
@@ -1262,7 +1262,36 @@ public extension Array where Element: Differentiable {
   }
 
   @usableFromInline
-  @derivative(of: differentiableReduce)
+  @derivative(of: differentiableReduce, wrt: (self, initialResult))
+  func _jvpDifferentiableReduce<Result: Differentiable>(
+    _ initialResult: Result,
+    _ nextPartialResult: @differentiable (Result, Element) -> Result
+  ) -> (value: Result,
+        differential: (Array.TangentVector, Result.TangentVector)
```
Comment (dan-zheng) [inline stdlib/public/Differentiation/DifferentiationSupport.swift]: minor:
```suggestion
        differential: (TangentVector, Result.TangentVector)
```

**Review by dan-zheng — APPROVED** [review]
LGTM with tests

File: stdlib/public/Differentiation/DifferentiationSupport.swift (hunk 1 above)
Comment (rxwei) [reply stdlib/public/Differentiation/DifferentiationSupport.swift]: I think keeping this is good for clarity. The VJP also keeps `Array.`. I don't feel strongly though.

**Comment by rxwei** [comment]
I currently don't have a bandwidth to write unit tests. This is just upstreaming parts of [this gist](https://gist.github.com/rxwei/edcbc7c06cff8d8fa47806974cdd976e). If anyone wants to commit some unit tests that'd be appreciated!

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by dan-zheng** [comment]
I accidentally deleted `tensorflow` branch, which closed this PR. That was not intentional, sorry!
It would be nice to protect `tensorflow` branch against deletion to prevent this from happening again.

**Comment by dan-zheng** [comment]
This PR was effectively cherry-picked into https://github.com/apple/swift/pull/32840. Thanks @rxwei for the original implementation!

**Pull request closed without merging**